File: basic_dsa_one_hundred_ninety_seven.java
/**
 * Basic DSA Problems - Longest Substring Without Repeating Characters
 * Length of longest substring with all distinct characters
 */
import java.util.HashMap;
import java.util.Map;

public class basic_dsa_one_hundred_ninety_seven {

    // Problem: String s. Return length of longest substring without repeating characters.

    /**
     * Sliding window + map (char -> last index). Extend right; if char seen, move left past it.
     * Time O(n), Space O(min(n, alphabet))
     */
    public static int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> idx = new HashMap<>();
        int max = 0, left = 0;
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            if (idx.containsKey(c) && idx.get(c) >= left)
                left = idx.get(c) + 1;
            idx.put(c, right);
            max = Math.max(max, right - left + 1);
        }
        return max;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Longest Substring Without Repeating Characters");
        System.out.println("=".repeat(60));

        System.out.println("abcabcbb -> " + lengthOfLongestSubstring("abcabcbb"));
        System.out.println("bbbbb -> " + lengthOfLongestSubstring("bbbbb"));
        System.out.println("pwwkew -> " + lengthOfLongestSubstring("pwwkew"));
    }
}

File: basic_dsa_one_hundred_ninety_seven.py
"""
Basic DSA - Roman to Integer
Convert a Roman numeral to an integer.
"""


def roman_to_int(s):
    """
    Scan left-to-right; if current value < next value, subtract, else add.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    val = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
    total = 0
    for i, ch in enumerate(s):
        cur = val[ch]
        nxt = val[s[i + 1]] if i + 1 < len(s) else 0
        total += -cur if cur < nxt else cur
    return total


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Roman to Integer")
    print("=" * 60)

    print(roman_to_int("III"))  # 3
    print(roman_to_int("LVIII"))  # 58
    print(roman_to_int("MCMXCIV"))  # 1994
